File: benchmarks/bench_closures.js
#!/usr/bin/env node
// Node counterpart of bench_closures.strada (identical workload).
function makeAdder(base) { const bias = base * 2; return (x) => x + bias; }
function makeOuter(seed) {
    const level0 = seed;
    const mid = () => (x) => x + level0;
    return mid();
}

let t0 = Date.now();
let made = 0;
for (let i = 0; i < 500000; i++) { const f = makeAdder(i); made++; }
let t1 = Date.now();
console.log("create:", made, (t1 - t0) / 1000);

const add7 = makeAdder(7);
let sum = 0;
for (let i = 0; i < 5000000; i++) sum += add7(i % 100);
let t2 = Date.now();
console.log("invoke:", sum, (t2 - t1) / 1000);

let counter = 0;
const bump = () => ++counter;
for (let i = 0; i < 2000000; i++) bump();
let t3 = Date.now();
console.log("capture-rw:", counter, (t3 - t2) / 1000);

const table = {};
for (let i = 0; i < 16; i++) table["op" + i] = makeAdder(i);
let tsum = 0;
for (let i = 0; i < 200000; i++) tsum += table["op" + (i % 16)](i % 50);
let t4 = Date.now();
console.log("table:", tsum, (t4 - t3) / 1000);

const deep = makeOuter(11);
let dsum = 0;
for (let i = 0; i < 200000; i++) dsum += deep(i % 10);
let t5 = Date.now();
console.log("transitive:", dsum, (t5 - t4) / 1000);
console.log("total:", (t5 - t0) / 1000);

File: benchmarks/bench_closures.pl
#!/usr/bin/env perl
# Perl counterpart of bench_closures.strada (identical workload).
use strict; use warnings; use Time::HiRes qw(time);

sub make_adder { my $base = shift; my $bias = $base * 2; return sub { $_[0] + $bias } }
sub make_outer {
    my $seed = shift;
    my $level0 = $seed;
    my $mid = sub { return sub { $_[0] + $level0 } };
    return $mid->();
}

my $t0 = time;
my $made = 0;
for my $i (0..499999) { my $f = make_adder($i); $made++ }
my $t1 = time;
printf "create: %d %.6f\n", $made, $t1 - $t0;

my $add7 = make_adder(7);
my $sum = 0;
$sum += $add7->($_ % 100) for 0..4999999;
my $t2 = time;
printf "invoke: %d %.6f\n", $sum, $t2 - $t1;

my $counter = 0;
my $bump = sub { return ++$counter };
$bump->() for 1..2000000;
my $t3 = time;
printf "capture-rw: %d %.6f\n", $counter, $t3 - $t2;

my %table = map { ("op$_" => make_adder($_)) } 0..15;
my $tsum = 0;
for my $i (0..199999) { $tsum += $table{"op" . ($i % 16)}->($i % 50) }
my $t4 = time;
printf "table: %d %.6f\n", $tsum, $t4 - $t3;

my $deep = make_outer(11);
my $dsum = 0;
$dsum += $deep->($_ % 10) for 0..199999;
my $t5 = time;
printf "transitive: %d %.6f\n", $dsum, $t5 - $t4;
printf "total: %.6f\n", $t5 - $t0;
